File: sdk/python/dynaep/temporal/validator.py
# ===========================================================================
# dynaep.temporal.validator - Temporal Validator
# Validates AG-UI event timestamps against the BridgeClock, detecting drift,
# future timestamps, stale events, and causal ordering violations.
# Every event passing through validation receives a BridgeTimestamp annotation.
# ===========================================================================

from __future__ import annotations
import time
import logging
from dataclasses import dataclass, field
from typing import Optional

from dynaep.temporal.clock import BridgeClock, BridgeTimestamp

logger = logging.getLogger("dynaep.temporal.validator")

# ---------------------------------------------------------------------------
# Maximum number of rejection log entries to retain in memory
# ---------------------------------------------------------------------------
MAX_REJECTION_LOG_SIZE = 500


# ---------------------------------------------------------------------------
# Types
# ---------------------------------------------------------------------------


@dataclass
class TemporalViolation:
    """Describes a single temporal constraint violation."""
    type: str  # "drift_exceeded" | "future_timestamp" | "stale_event" | "causal_violation"
    detail: str
    agent_time_ms: Optional[int]
    bridge_time_ms: int
    threshold_ms: int


@dataclass
class TemporalValidationResult:
    """Result of validating a single event through the temporal pipeline."""
    accepted: bool
    bridge_timestamp: BridgeTimestamp
    violations: list
    warnings: list


@dataclass
class TemporalValidatorConfig:
    """Configuration controlling the strictness of temporal validation."""
    max_drift_ms: int = 50
    max_future_ms: int = 500
    max_staleness_ms: int = 5000
    overwrite_timestamps: bool = True
    log_rejections: bool = True
    mode: str = "strict"  # "strict" | "permissive" | "log_only"


# ---------------------------------------------------------------------------
# Internal bookkeeping
# ---------------------------------------------------------------------------


@dataclass
class _RejectionLogEntry:
    """Internal record of a rejected or violated event."""
    event_type: str
    violations: list
    recorded_at: int


# ---------------------------------------------------------------------------
# TemporalValidator
# ---------------------------------------------------------------------------


class TemporalValidator:
    """Validates AG-UI event timestamps against the BridgeClock.

    Detects drift, future timestamps, and stale events. Operates in three
    modes: strict (reject on violation), permissive (accept with warnings),
    and log_only (accept all, log violations silently).
    """

    def __init__(self, clock: BridgeClock, config: TemporalValidatorConfig) -> None:
        self._clock = clock
        self._config = config
        self._rejection_log: list[_RejectionLogEntry] = []
        self._last_validated_bridge_ms: int = 0

    # -----------------------------------------------------------------------
    # Primary validation entry point
    # -----------------------------------------------------------------------

    def validate(self, event: dict) -> TemporalValidationResult:
        """Run the full temporal validation pipeline on a single event.

        Steps:
        1. Read the agent-supplied timestamp (may be absent)
        2. Produce the authoritative BridgeTimestamp from the clock
        3. Overwrite the event timestamp when configured
        4. Check drift between agent time and bridge time
        5. Check for future timestamps
        6. Check for staleness
        7. Attach BridgeTimestamp metadata to the event
        8. Determine acceptance based on mode
        """
        violations: list[TemporalViolation] = []
        warnings: list[str] = []

        # Step 1: Read the agent-supplied timestamp
        raw_ts = event.get("timestamp")
        agent_time_ms: Optional[int] = int(raw_ts) if isinstance(raw_ts, (int, float)) else None

        # Step 2: Produce the authoritative BridgeTimestamp
        bridge_timestamp = self._clock.stamp(agent_time_ms)
        bridge_time_ms = bridge_timestamp.bridge_time_ms

        # Step 3: Overwrite the event timestamp when configured
        if self._config.overwrite_timestamps:
            event["timestamp"] = bridge_time_ms
            warnings.append(
                "Agent timestamp overwritten with bridge time " + str(bridge_time_ms)
            )

        # Step 4: Check drift
        if agent_time_ms is not None:
            drift_ok, drift_ms = self.check_drift(agent_time_ms)
            if not drift_ok:
                violations.append(TemporalViolation(
                    type="drift_exceeded",
                    detail=(
                        "Drift of " + str(drift_ms) + "ms exceeds maximum allowed "
                        + str(self._config.max_drift_ms) + "ms"
                    ),
                    agent_time_ms=agent_time_ms,
                    bridge_time_ms=bridge_time_ms,
                    threshold_ms=self._config.max_drift_ms,
                ))

        # Step 5: Check for future timestamps
        if agent_time_ms is not None:
            is_future = self.check_future_timestamp(agent_time_ms)
            if is_future:
                violations.append(TemporalViolation(
                    type="future_timestamp",
                    detail=(
                        "Agent timestamp " + str(agent_time_ms)
                        + "ms is beyond the allowed future window of "
                        + str(self._config.max_future_ms) + "ms past bridge time"
                    ),
                    agent_time_ms=agent_time_ms,
                    bridge_time_ms=bridge_time_ms,
                    threshold_ms=self._config.max_future_ms,
                ))

        # Step 6: Check for staleness
        if agent_time_ms is not None:
            is_stale = self.check_staleness(agent_time_ms)
            if is_stale:
                violations.append(TemporalViolation(
                    type="stale_event",
                    detail=(
                        "Event is " + str(bridge_time_ms - agent_time_ms)
                        + "ms old, exceeding staleness limit of "
                        + str(self._config.max_staleness_ms) + "ms"
                    ),
                    agent_time_ms=agent_time_ms,
                    bridge_time_ms=bridge_time_ms,
                    threshold_ms=self._config.max_staleness_ms,
                ))

        # Step 7: Attach BridgeTimestamp metadata to the event
        event["_temporal"] = {
            "bridge_time_ms": bridge_timestamp.bridge_time_ms,
            "drift_ms": bridge_timestamp.drift_ms,
            "source": bridge_timestamp.source,
            "validated_at": int(time.time() * 1000),
        }

        # Update tracking for causal ordering checks
        self._last_validated_bridge_ms = bridge_time_ms

        # Step 8: Determine acceptance based on mode
        accepted = self._resolve_acceptance(violations, warnings, event)

        return TemporalValidationResult(
            accepted=accepted,
            bridge_timestamp=bridge_timestamp,
            violations=violations,
            warnings=warnings,
        )

    # -----------------------------------------------------------------------
    # Individual checks
    # -----------------------------------------------------------------------

    def check_drift(self, agent_time_ms: int) -> tuple:
        """Compare agent time against current bridge time.

        Returns a tuple of (within_tolerance: bool, drift_ms: int).
        The drift is the absolute difference between the two clocks.
        """
        current_bridge_ms = self._clock.now()
        drift_ms = abs(agent_time_ms - current_bridge_ms)
        within_tolerance = drift_ms <= self._config.max_drift_ms
        return (within_tolerance, drift_ms)

    def check_future_timestamp(self, agent_time_ms: int) -> bool:
        """Check if the agent timestamp is too far ahead of bridge time.

        Returns True if the agent time exceeds the bridge time by more
        than the configured max_future_ms window.
        """
        current_bridge_ms = self._clock.now()
        ahead_ms = agent_time_ms - current_bridge_ms
        exceeds_future_window = ahead_ms > self._config.max_future_ms
        return exceeds_future_window

    def check_staleness(self, agent_time_ms: int) -> bool:
        """Check if the event is too old relative to bridge time.

        Returns True if the age of the event exceeds the configured
        max_staleness_ms threshold.
        """
        current_bridge_ms = self._clock.now()
        age_ms = current_bridge_ms - agent_time_ms
        is_beyond_limit = age_ms > self._config.max_staleness_ms
        return is_beyond_limit

    # -----------------------------------------------------------------------
    # Batch validation
    # -----------------------------------------------------------------------

    def validate_batch(self, events: list) -> list:
        """Validate a list of events in order, returning a result for each.

        Events are processed sequentially so that each event's validation
        can observe the side effects (e.g. timestamp overwrite) of the
        events before it.
        """
        results: list[TemporalValidationResult] = []
        batch_size = len(events)
        for idx in range(batch_size):
            current_event = events[idx]
            result = self.validate(current_event)
            results.append(result)
        return results

    # -----------------------------------------------------------------------
    # Internal helpers
    # -----------------------------------------------------------------------

    def _resolve_acceptance(
        self,
        violations: list[TemporalViolation],
        warnings: list[str],
        event: dict,
    ) -> bool:
        """Determine whether the event should be accepted based on mode.

        In strict mode, any violation causes rejection.
        In permissive mode, violations become warnings but the event is accepted.
        In log_only mode, everything is accepted and violations are logged silently.
        """
        mode = self._config.mode
        has_violations = len(violations) > 0

        if mode == "strict":
            if has_violations and self._config.log_rejections:
                self._record_rejection(event, violations)
            return not has_violations

        if mode == "permissive":
            if has_violations:
                for v in violations:
                    warnings.append("[permissive] " + v.type + ": " + v.detail)
                if self._config.log_rejections:
                    self._record_rejection(event, violations)
            return True

        # log_only mode: accept everything, log all violations silently
        if has_violations:
            self._record_rejection(event, violations)
            for v in violations:
                warnings.append("[log_only] " + v.type + ": " + v.detail)
        return True

    def _record_rejection(
        self,
        event: dict,
        violations: list[TemporalViolation],
    ) -> None:
        """Record a rejection to the in-memory log.

        The log is capped at MAX_REJECTION_LOG_SIZE entries to prevent
        unbounded memory growth. Oldest entries are evicted first.
        """
        entry = _RejectionLogEntry(
            event_type=event.get("type", "unknown"),
            violations=list(violations),
            recorded_at=int(time.time() * 1000),
        )
        self._rejection_log.append(entry)

        if len(self._rejection_log) > MAX_REJECTION_LOG_SIZE:
            excess = len(self._rejection_log) - MAX_REJECTION_LOG_SIZE
            self._rejection_log = self._rejection_log[excess:]

File: sdk/typescript/src/temporal/validator.ts
// ===========================================================================
// @dynaep/temporal - Temporal Validator
// Validates AG-UI event timestamps against the BridgeClock, detecting drift,
// future timestamps, stale events, and causal ordering violations.
// Every event passing through validation receives a BridgeTimestamp annotation.
// ===========================================================================

import type { BridgeClock, BridgeTimestamp } from "./clock";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface TemporalViolation {
  type: "drift_exceeded" | "future_timestamp" | "stale_event" | "causal_violation";
  detail: string;
  agentTimeMs: number | null;
  bridgeTimeMs: number;
  thresholdMs: number;
}

export interface TemporalValidationResult {
  accepted: boolean;
  bridgeTimestamp: BridgeTimestamp;
  violations: TemporalViolation[];
  warnings: string[];
}

export interface TemporalValidatorConfig {
  maxDriftMs: number;
  maxFutureMs: number;
  maxStalenessMs: number;
  overwriteTimestamps: boolean;
  logRejections: boolean;
  mode: "strict" | "permissive" | "log_only";
}

interface AGUIEvent {
  type: string;
  timestamp?: number;
  dynaep_type?: string;
  target_id?: string;
  _temporal?: any;
  [key: string]: any;
}

// ---------------------------------------------------------------------------
// Rejection Log Entry (internal bookkeeping)
// ---------------------------------------------------------------------------

interface RejectionLogEntry {
  eventType: string;
  violations: TemporalViolation[];
  recordedAt: number;
}

// ---------------------------------------------------------------------------
// TemporalValidator
// ---------------------------------------------------------------------------

export class TemporalValidator {
  private readonly clock: BridgeClock;
  private readonly config: TemporalValidatorConfig;
  private readonly rejectionLog: RejectionLogEntry[];
  private lastValidatedBridgeMs: number;

  constructor(clock: BridgeClock, config: TemporalValidatorConfig) {
    this.clock = clock;
    this.config = config;
    this.rejectionLog = [];
    this.lastValidatedBridgeMs = 0;
  }

  // -------------------------------------------------------------------------
  // Primary validation entry point
  // -------------------------------------------------------------------------

  validate(event: AGUIEvent): TemporalValidationResult {
    const violations: TemporalViolation[] = [];
    const warnings: string[] = [];

    // Step 1: Read the agent-supplied timestamp (may be absent)
    const agentTimeMs: number | null =
      typeof event.timestamp === "number" ? event.timestamp : null;

    // Step 2: Produce the authoritative BridgeTimestamp from the clock
    const bridgeTimestamp: BridgeTimestamp = this.clock.stamp(
      agentTimeMs !== null ? agentTimeMs : undefined,
    );
    const bridgeTimeMs: number = bridgeTimestamp.bridgeTimeMs;

    // Step 3: Overwrite the event timestamp when configured to do so
    if (this.config.overwriteTimestamps) {
      event.timestamp = bridgeTimeMs;
      warnings.push(
        "Agent timestamp overwritten with bridge time " + String(bridgeTimeMs),
      );
    }

    // Step 4: Check drift between agent time and bridge time
    if (agentTimeMs !== null) {
      const driftResult = this.checkDrift(agentTimeMs);
      if (!driftResult.withinTolerance) {
        const driftViolation: TemporalViolation = {
          type: "drift_exceeded",
          detail:
            "Drift of " +
            String(driftResult.driftMs) +
            "ms exceeds maximum allowed " +
            String(this.config.maxDriftMs) +
            "ms",
          agentTimeMs,
          bridgeTimeMs,
          thresholdMs: this.config.maxDriftMs,
        };
        violations.push(driftViolation);
      }
    }

    // Step 5: Check for future timestamps
    if (agentTimeMs !== null) {
      const isFuture = this.checkFutureTimestamp(agentTimeMs);
      if (isFuture) {
        const futureViolation: TemporalViolation = {
          type: "future_timestamp",
          detail:
            "Agent timestamp " +
            String(agentTimeMs) +
            "ms is beyond the allowed future window of " +
            String(this.config.maxFutureMs) +
            "ms past bridge time",
          agentTimeMs,
          bridgeTimeMs,
          thresholdMs: this.config.maxFutureMs,
        };
        violations.push(futureViolation);
      }
    }

    // Step 6: Check for staleness
    if (agentTimeMs !== null) {
      const isStale = this.checkStaleness(agentTimeMs);
      if (isStale) {
        const staleViolation: TemporalViolation = {
          type: "stale_event",
          detail:
            "Event is " +
            String(bridgeTimeMs - agentTimeMs) +
            "ms old, exceeding staleness limit of " +
            String(this.config.maxStalenessMs) +
            "ms",
          agentTimeMs,
          bridgeTimeMs,
          thresholdMs: this.config.maxStalenessMs,
        };
        violations.push(staleViolation);
      }
    }

    // Step 7: Attach BridgeTimestamp metadata to the event
    event._temporal = {
      bridgeTimeMs: bridgeTimestamp.bridgeTimeMs,
      driftMs: bridgeTimestamp.driftMs,
      source: bridgeTimestamp.source,
      validatedAt: Date.now(),
    };

    // Update tracking for causal ordering checks
    this.lastValidatedBridgeMs = bridgeTimeMs;

    // Step 8-10: Determine acceptance based on the configured mode
    const accepted = this.resolveAcceptance(violations, warnings, event);

    const result: TemporalValidationResult = {
      accepted,
      bridgeTimestamp,
      violations,
      warnings,
    };

    return result;
  }

  // -------------------------------------------------------------------------
  // Drift check - compares agent time against current bridge time
  // -------------------------------------------------------------------------

  checkDrift(agentTimeMs: number): { withinTolerance: boolean; driftMs: number } {
    const currentBridgeMs: number = this.clock.now();
    const driftMs: number = Math.abs(agentTimeMs - currentBridgeMs);
    const withinTolerance: boolean = driftMs <= this.config.maxDriftMs;
    return { withinTolerance, driftMs };
  }

  // -------------------------------------------------------------------------
  // Future timestamp check - true if the agent time is too far ahead
  // -------------------------------------------------------------------------

  checkFutureTimestamp(agentTimeMs: number): boolean {
    const currentBridgeMs: number = this.clock.now();
    const aheadMs: number = agentTimeMs - currentBridgeMs;
    const exceedsFutureWindow: boolean = aheadMs > this.config.maxFutureMs;
    return exceedsFutureWindow;
  }

  // -------------------------------------------------------------------------
  // Staleness check - true if the event is too old relative to bridge time
  // -------------------------------------------------------------------------

  checkStaleness(agentTimeMs: number): boolean {
    const currentBridgeMs: number = this.clock.now();
    const ageMs: number = currentBridgeMs - agentTimeMs;
    const isBeyondStalenessLimit: boolean = ageMs > this.config.maxStalenessMs;
    return isBeyondStalenessLimit;
  }

  // -------------------------------------------------------------------------
  // Batch validation - processes each event in order
  // -------------------------------------------------------------------------

  validateBatch(events: AGUIEvent[]): TemporalValidationResult[] {
    const results: TemporalValidationResult[] = [];
    const batchSize: number = events.length;

    for (let idx = 0; idx < batchSize; idx++) {
      const currentEvent: AGUIEvent = events[idx];
      const result: TemporalValidationResult = this.validate(currentEvent);
      results.push(result);
    }

    return results;
  }

  // -------------------------------------------------------------------------
  // Internal: resolve acceptance based on mode and violation state
  // -------------------------------------------------------------------------

  private resolveAcceptance(
    violations: TemporalViolation[],
    warnings: string[],
    event: AGUIEvent,
  ): boolean {
    const mode = this.config.mode;
    const hasViolations = violations.length > 0;

    if (mode === "strict") {
      // In strict mode, any violation results in rejection
      if (hasViolations && this.config.logRejections) {
        this.recordRejection(event, violations);
      }
      return !hasViolations;
    }

    if (mode === "permissive") {
      // In permissive mode, we always accept but record violations as warnings
      if (hasViolations) {
        for (const v of violations) {
          warnings.push("[permissive] " + v.type + ": " + v.detail);
        }
        if (this.config.logRejections) {
          this.recordRejection(event, violations);
        }
      }
      return true;
    }

    // log_only mode: accept everything, log all violations silently
    if (hasViolations) {
      this.recordRejection(event, violations);
      for (const v of violations) {
        warnings.push("[log_only] " + v.type + ": " + v.detail);
      }
    }
    return true;
  }

  // -------------------------------------------------------------------------
  // Internal: record a rejection to the in-memory log
  // -------------------------------------------------------------------------

  private recordRejection(
    event: AGUIEvent,
    violations: TemporalViolation[],
  ): void {
    const entry: RejectionLogEntry = {
      eventType: event.type || "unknown",
      violations: violations.slice(),
      recordedAt: Date.now(),
    };
    this.rejectionLog.push(entry);

    // Cap the log at 500 entries to prevent unbounded growth
    if (this.rejectionLog.length > 500) {
      this.rejectionLog.splice(0, this.rejectionLog.length - 500);
    }
  }
}
